### PR TITLE
1 implement batch processing of nodes

### PR DESCRIPTION
**Pull request by catorghans**
Batch processing pull request.

### CODE REVIEW

File: src/Form/CoyoteImgBatchForm.php [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  public function getFormId() {
+    return $this->t('coyote_img_batch_form');
+  }
+
+  private static function isDefined(?string $var): bool
+  {
+    return !is_null($var) && strlen($var) > 0;
+  }
+
+  private function getProfileRole(?ProfileModel $profile): string
```
Comment (jkva) [inline src/Form/CoyoteImgBatchForm.php:33]: The signature is `string` but it returns `null` in some cases.

File: src/Form/CoyoteImgBatchForm.php [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+   */
+  public function buildForm(array $form, FormStateInterface $form_state): array {
+    $config = $this->config('coyote_img_desc.settings');
+    $token = $config->get('api_token');
+    $endpoint = $config->get('api_endpoint');
+    $resourceGroup = $config->get('api_resource_group');
+    $organizationId = $config->get('api_organization');
+    if (self::isDefined($token) && self::isDefined($endpoint)) {
+      $profile = CoyoteApiClientHelperFunctions::getProfile($endpoint, $token);
+      $role = self::getProfileRole($profile);
```
Comment (jkva) [inline src/Form/CoyoteImgBatchForm.php:59]: `getProfileRole` is called as static, but uses `$this` in its implementation.

File: src/Form/CoyoteImgBatchForm.php [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+     }
+
+     $results[] = $nid;
+
+     $context['message'] = $message;
+     $context['results'] = $results;
+
+  }
+
+  public function batch_processing_nodes_finished($success, $results, $operations) {
```
Comment (jkva) [inline src/Form/CoyoteImgBatchForm.php:139]: Isn't the calling of this function disabled on line 98?

File: src/Form/CoyoteImgBatchForm.php [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  public function submitForm(array &$form, FormStateInterface $form_state): void {
+     $nids = \Drupal::entityQuery('node')->execute();
+
+     foreach ($nids as $nid){     
+       $operations[] = [['\Drupal\coyote_img_desc\Form\CoyoteImgBatchForm','batch_processing_nodes'], [$nid]];
+     }
+     $batch = [
+          'title' => $this->t('Batch processing All Nodes ...'),
+          'operations' => $operations,
+        //  'finished' => ['Drupal\coyote_img_desc\Form\CoyoteImgBatchForm','batch_processing_nodes_finished'],
```
Comment (jkva) [inline src/Form/CoyoteImgBatchForm.php:98]: Dead code?

File: src/Form/CoyoteImgBatchForm.php [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    $endpoint = $config->get('api_endpoint');
+    $resourceGroup = $config->get('api_resource_group');
+    $organizationId = $config->get('api_organization');
+    if (self::isDefined($token) && self::isDefined($endpoint)) {
+      $profile = CoyoteApiClientHelperFunctions::getProfile($endpoint, $token);
+      $role = self::getProfileRole($profile);
+    }
+     
+    $disabled = false;
+    if (!self::isDefined($token) || !self::isDefined($endpoint) || !self::isDefined($organizationId) || !self::isDefined($resourceGroup) || is_null($profile) || is_null($role)) {
```
Comment (jkva) [inline src/Form/CoyoteImgBatchForm.php:63]: Might result in more intuitive code to abstract a long chain of `or` into its own function such as `private function hasValidBatchProcessingConfig(): bool` or something like that.

File: src/Form/CoyoteImgBatchForm.php [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+     }
+     $batch = [
+          'title' => $this->t('Batch processing All Nodes ...'),
+          'operations' => $operations,
+        //  'finished' => ['Drupal\coyote_img_desc\Form\CoyoteImgBatchForm','batch_processing_nodes_finished'],
+      ];
+      batch_set($batch);
+  }
+  
+  public function batch_processing_nodes($nid, &$context) {
```
Comment (jkva) [inline src/Form/CoyoteImgBatchForm.php:103]: Inconsistent casing in function name.

File: src/Form/CoyoteImgBatchForm.php [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+     $message = 'Proccessing ALL nodes to coyote';
+
+     $results = array();
```
Comment (jkva) [inline src/Form/CoyoteImgBatchForm.php:106]: Can we consistently use `[]` shorthand?

File: src/Form/CoyoteImgBatchForm.php [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+     $isPublished = $node->isPublished();
+     $coyoteProcessUnpublishedNodes = $config->get('coyote_process_unpublished_nodes');
+
+     if (!$coyoteProcessUnpublishedNodes && !$isPublished) return false;
+     $contentHelper = new ContentHelper($output);
+
+     $images = $contentHelper->getImages();
+     foreach($images as $image) {
+        $resource = Util::getImageResource($image, $hostUri);
```
Comment (jkva) [inline src/Form/CoyoteImgBatchForm.php:129]: `$resource` isn't used anywhere, we probably want to track (log?) if this is `null`.

File: coyote_img_desc.module [hunk 9]
```diff
@@ -10,4 +10,4 @@ function coyote_img_desc_node_view_alter(array &$build, EntityInterface $entity,
 {
   $build['_coyote_node_url'] = $entity->toUrl('canonical', ['absolute' => true])->toString();
   $build['#post_render'][] = [NodeViewAlterPostRenderHook::class, 'hook'];
-}
\ No newline at end of file
+}
```
Comment (jkva) [inline coyote_img_desc.module:13]: Seems only a whitespace change.

**Review by jkva — CHANGES_REQUESTED** [review]
I've left some comments. Once those are addressed I'll run some tests with it and if that's all good this will be a welcome added feature 👍 

There are quite a lot of commits for not a very complex PR, consider converting the final PR into a small amount of descriptive atomic commits via a soft reset to `origin/main`.

File: src/Form/CoyoteImgBatchForm.php (hunk 8 above)
Comment (catorghans) [reply src/Form/CoyoteImgBatchForm.php]: one extra commit. Please have a look.

File: src/Form/CoyoteImgBatchForm.php [hunk 10]
```diff
@@ -30,7 +30,7 @@ private static function isDefined(?string $var): bool
     return !is_null($var) && strlen($var) > 0;
   }
 
-  private function getProfileRole(?ProfileModel $profile): string
+  private function getProfileRole(?ProfileModel $profile)
```
Comment (jkva) [inline src/Form/CoyoteImgBatchForm.php:33]: How about a signature like `?string`?

**Review by jkva — CHANGES_REQUESTED** [review]
One signature nitpick here. Can you check back to my earlier comment about rebasing? Thanks 👍

**Comment by jkva** [comment]
I'm trying to test the functionality, but get an error while running on localhost. Could that be related?

```Error: Call to a member function claimItem() on null in /var/www/html/web/core/includes/batch.inc on line 285 #0 /var/www/html/web/core/includes/batch.inc(137): _batch_process()\n#1 [...]```

and

``` Error: Call to a member function getAllItems() on null in /var/www/html/web/core/includes/batch.inc on line 455 #0 /var/www/html/web/core/includes/batch.inc(98): _batch_finished()\n#1 /var/www/html/web/core/modules/system/src/Controller/BatchController.php(55): _batch_page() [...]```

**Comment by jkva** [comment]
> Probably related. So how does your localhost site look? Do you only have nodes without images? Or no nodes at all? (am trying to figure out what the null could be)

This is a fresh install, so likely the case of no nodes existing. Is that easy to guard against?

**Comment by catorghans** [comment]
> > Probably related. So how does your localhost site look? Do you only have nodes without images? Or no nodes at all? (am trying to figure out what the null could be)
> 
> This is a fresh install, so likely the case of no nodes existing. Is that easy to guard against?

No nodes at all! Ok, had not tested that. I can confirm the issue. Solution will follow.

**Comment by catorghans** [comment]
> > > Probably related. So how does your localhost site look? Do you only have nodes without images? Or no nodes at all? (am trying to figure out what the null could be)
> > 
> > 
> > This is a fresh install, so likely the case of no nodes existing. Is that easy to guard against?
> 
> No nodes at all! Ok, had not tested that. I can confirm the issue. Solution will follow.

Try again with latest commit

**Comment by jkva** [comment]
That fixed it 👍

**Review by jkva — APPROVED** [review]

**Pull request merged**